File: Assets/src/model/tasks/SleepTask.cs

using System;
using System.Collections.Generic;
using System.Linq;
/// Don't do anything until the Player's in view
[System.Serializable]
class SleepTask : ActorTask, IAttackDamageTakenModifier, ITakeAnyDamageHandler {
  public override TaskStage WhenToCheckIsDone => TaskStage.After;
  private bool done = false;
  private int? maxTurns;
  public readonly bool isDeepSleep;

  public bool wakeUpNextTurn { get; set; }
  public SleepTask(Actor actor, int? maxTurns = null, bool isDeepSleep = false) : base(actor) {
    this.maxTurns = maxTurns;
    this.isDeepSleep = isDeepSleep;
  }

  /// doubles attack damage taken while sleeping
  public int Modify(int input) {
    return input * 2;
  }

  /// wake up when hurt!
  public void HandleTakeAnyDamage(int damage) {
    if (damage > 0) {
      // Wake up early by ending this task immediately (will trigger Ended())
      actor.statuses.Add(new SurprisedStatus());
      actor.GoToNextTask();
    }
  }

  protected virtual bool ShouldWakeUp() {
    if (wakeUpNextTurn) {
      return true;
    }

    if (maxTurns != null && maxTurns <= 0) {
      return true;
    }

    if (isDeepSleep) {
      return false;
    }

    return actor.isVisible && actor.CanTargetPlayer();
  }

  protected override BaseAction GetNextActionImpl() {
    if (maxTurns != null) {
      maxTurns--;
    }
    if (ShouldWakeUp()) {
      // // end of sleep - wake up adjacent sleeping Actors
      // foreach (var actor in actor.floor.AdjacentActors(actor.pos)) {
      //   if (actor.task is SleepTask s && !s.isDeepSleep) {
      //     // hack to wake them up
      //     s.wakeUpNextTurn = true;
      //   }
      // }
      done = true;
      actor.statuses.Add(new SurprisedStatus());
    }
    return new WaitBaseAction(actor);
  }

  public override bool IsDone() => done;
}

[System.Serializable]
[ObjectInfo("colored_transparent_packed_658", description: "You're surprised! You must spend the next turn shaking it off.")]
class SurprisedStatus : Status, IBaseActionModifier {
  public override bool isDebuff => true;
  public override string Info() => "";
  private bool remove = false;

  public BaseAction Modify(BaseAction input) {
    /// The SurprisedStatus only affects one action but it ends at the start of the second action,
    /// so the player can see the (!)
    if (remove) {
      Remove();
      return input;
    } else {
      // treat player differently - remove status immediately, and do a struggle
      if (actor == GameModel.main.player) {
        Remove();
        return new StruggleBaseAction(actor);
      } else {
        remove = true;
        return new WaitBaseAction(input.actor);
      }
    }
  }

  public override bool Consume(Status other) => true;
}

File: web/src/model/tasks/SleepTask.ts
import { ActorTask, TaskStage } from '../ActorTask';
import { BaseAction, WaitBaseAction, StruggleBaseAction } from '../BaseAction';
import { Status } from '../Status';
import { GameModelRef } from '../GameModelRef';
import {
  ATTACK_DAMAGE_TAKEN_MOD,
  BASE_ACTION_MOD,
} from '../../core/Modifiers';
import { TAKE_ANY_DAMAGE_HANDLER } from '../Body';
import type { Actor } from '../Actor';

export class SleepTask extends ActorTask {
  private done = false;
  private maxTurns: number | null;
  readonly isDeepSleep: boolean;
  wakeUpNextTurn = false;

  // Modifier/handler symbols
  readonly [ATTACK_DAMAGE_TAKEN_MOD] = true;
  readonly [TAKE_ANY_DAMAGE_HANDLER] = true;

  get whenToCheckIsDone(): TaskStage {
    return TaskStage.After;
  }

  get isPlayerOverridable(): boolean {
    return this.isDeepSleep;
  }

  constructor(actor: Actor, maxTurns: number | null = null, isDeepSleep = false) {
    super(actor);
    this.maxTurns = maxTurns;
    this.isDeepSleep = isDeepSleep;
  }

  /** IAttackDamageTakenModifier — doubles attack damage while sleeping */
  modify(input: number): number {
    return input * 2;
  }

  /** ITakeAnyDamageHandler — wake up when hurt */
  handleTakeAnyDamage(damage: number): void {
    if (damage > 0) {
      this.actor.statuses.add(new SurprisedStatus());
      this.actor.goToNextTask();
    }
  }

  protected shouldWakeUp(): boolean {
    if (this.wakeUpNextTurn) return true;
    if (this.maxTurns !== null && this.maxTurns <= 0) return true;
    if (this.isDeepSleep) return false;
    return this.actor.isVisible && this.actor.canTargetPlayer();
  }

  protected getNextActionImpl(): BaseAction {
    if (this.maxTurns !== null) this.maxTurns--;
    if (this.shouldWakeUp()) {
      this.done = true;
      this.actor.statuses.add(new SurprisedStatus());
    }
    return new WaitBaseAction(this.actor);
  }

  isDone(): boolean {
    return this.done;
  }
}

export class SurprisedStatus extends Status {
  readonly [BASE_ACTION_MOD] = true;
  private removeNext = false;

  get isDebuff(): boolean {
    return true;
  }

  /** Handles both STEP_MOD (inherited) and BASE_ACTION_MOD calls. */
  modify(input: any): any {
    if (!(input instanceof BaseAction)) {
      // STEP_MOD call — delegate to super (calls Step())
      return super.modify(input);
    }
    // BASE_ACTION_MOD call
    if (this.removeNext) {
      this.Remove();
      return input;
    }
    const player = GameModelRef.mainOrNull?.player;
    if (this.actor === player) {
      this.Remove();
      return new StruggleBaseAction(input.actor);
    }
    this.removeNext = true;
    return new WaitBaseAction(input.actor);
  }

  Consume(_other: Status): boolean {
    return true;
  }
}
